File: near/src/main/java/com/adroitandroid/near/discovery/server/UdpBroadcastListeningHandler.java
package com.adroitandroid.near.discovery.server;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.Message;

import androidx.collection.ArrayMap;

import com.adroitandroid.near.model.Host;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.SocketException;
import java.net.UnknownHostException;
import java.util.Set;

public class UdpBroadcastListeningHandler extends Handler {
    private static final int LISTEN = 5678;
    private static UdpBroadcastListeningHandler handler;
    private DatagramSocket mSocket;
    private UdpBroadcastListener mListener;
    private ArrayMap<Host, StaleHostHandler> mHostHandlerMap;
    private Set<String> mCurrentIps;
    private boolean isHostClientToo;
    private long mStaleTimeout;

    UdpBroadcastListeningHandler(Looper looper) {
        super(looper);
    }

    static void startBroadcastListening(final ArrayMap<Host, StaleHostHandler> hostHandlerMap,
                                        final Set<String> currentHostIps,
                                        final boolean isHostClientToo, final long staleTimeout) {
        HandlerThread handlerThread = new HandlerThread("ServerService") {
            @Override
            protected void onLooperPrepared() {
                handler = new UdpBroadcastListeningHandler(getLooper());
                handler.mHostHandlerMap = hostHandlerMap;
                handler.mCurrentIps = currentHostIps;
                handler.isHostClientToo = isHostClientToo;
                handler.mStaleTimeout = staleTimeout;
                handler.sendEmptyMessage(UdpBroadcastListeningHandler.LISTEN);
            }
        };
        handlerThread.start();
    }

    static void setListener(UdpBroadcastListener listener) {
        if (handler != null) {
            handler.updateListenersTo(listener);
        }
    }

    private void updateListenersTo(UdpBroadcastListener listener) {
        mListener = listener;
        synchronized (UdpBroadcastListeningHandler.this) {
            for (StaleHostHandler hostHandler : mHostHandlerMap.values()) {
                hostHandler.setListener(listener);
            }
        }
    }

    @Override
    public void handleMessage(Message msg) {
        if (msg.what == LISTEN) {
            try {
                if (mSocket == null) {
                    mSocket = new DatagramSocket(null);
                    InetSocketAddress socketAddress
                            = new InetSocketAddress(InetAddress.getByName("0.0.0.0"), 8888);
                    mSocket.setReuseAddress(true);
                    mSocket.setBroadcast(true);
                    mSocket.setSoTimeout(0);
                    mSocket.bind(socketAddress);
                }

                byte[] recvBuf = new byte[15000];
                DatagramPacket packet = new DatagramPacket(recvBuf, recvBuf.length);
                DatagramSocket socket = mSocket;
                socket.receive(packet);

                Host host = new Host(packet.getAddress(), new String(packet.getData()).trim());
                if (isHostClientToo || !mCurrentIps.contains(host.getHostAddress())) {
                    StaleHostHandler handler = mHostHandlerMap.get(host);
                    if (handler == null) {
                        handler = new StaleHostHandler(host, mHostHandlerMap, mListener);
                        synchronized (UdpBroadcastListeningHandler.this) {
                            mHostHandlerMap.put(host, handler);
                        }
                        if (mListener != null) {
                            new Handler(Looper.getMainLooper()).post(new Runnable() {
                                @Override
                                public void run() {
                                    mListener.onHostsUpdate(mHostHandlerMap.keySet());
                                }
                            });
                        }
                    } else if (hostNameChanged(host, mHostHandlerMap)) {
                        if (mListener != null) {
                            new Handler(Looper.getMainLooper()).post(new Runnable() {
                                @Override
                                public void run() {
                                    mListener.onHostsUpdate(mHostHandlerMap.keySet());
                                }
                            });
                        }
                    }
                    handler.removeMessages(StaleHostHandler.STALE_HOST);
                    handler.sendEmptyMessageDelayed(StaleHostHandler.STALE_HOST, mStaleTimeout);
                }
            } catch (SocketException | UnknownHostException e) {
                e.printStackTrace();
                if (mSocket != null) {
                    mSocket.close();
                    mSocket = null;
                }
            } catch (IOException e) {
                e.printStackTrace();
                if (mListener != null) {
                    new Handler(Looper.getMainLooper()).post(new Runnable() {
                        @Override
                        public void run() {
                            mListener.onReceiveFailed();
                        }
                    });
                }
            }
            if (getLooper().getThread().isAlive()) {
                sendEmptyMessage(LISTEN);
            }
        }
    }

    private boolean hostNameChanged(Host updatedHost, ArrayMap<Host, StaleHostHandler> hostHandlerMap) {
        for (Host host : hostHandlerMap.keySet()) {
            if (updatedHost.equals(host) && !updatedHost.getName().equals(host.getName())) {
                hostHandlerMap.put(updatedHost, hostHandlerMap.remove(host));
                return true;
            }
        }
        return false;
    }

    static void stopListeningForBroadcasts() {
        if (handler != null) {
            handler.stop();
            handler = null;
        }
    }

    private void stop() {
        for (StaleHostHandler handler : mHostHandlerMap.values()) {
            handler.removeMessages(StaleHostHandler.STALE_HOST);
        }
        mListener = null;
        mSocket.close();
        mSocket = null;
        removeMessages(UdpBroadcastListeningHandler.LISTEN);
        getLooper().quitSafely();
    }
}

File: near/src/main/java/com/adroitandroid/near/discovery/server/UdpBroadcastListeningHandler.kt
package com.adroitandroid.near.discovery.server

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.os.Message
import androidx.collection.ArrayMap
import com.adroitandroid.near.model.Host
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.*

class UdpBroadcastListeningHandler internal constructor(looper: Looper) : Handler(looper) {
    private var mSocket: DatagramSocket? = null
    private var mListener: UdpBroadcastListener? = null
    private lateinit var mHostHandlerMap: ArrayMap<Host, StaleHostHandler>
    private lateinit var mCurrentIps: Set<String>
    private var isHostClientToo = false
    private var mStaleTimeout: Long = 0
    private var mPort: Int = UdpServerService.DISCOVERY_PORT
    private var mRegex: Regex = Regex("^$")


    private fun updateListenersTo(listener: UdpBroadcastListener) {
        mListener = listener
        synchronized(this@UdpBroadcastListeningHandler) {
            for (hostHandler: StaleHostHandler in mHostHandlerMap.values) {
                hostHandler.setListener(listener)
            }
        }
    }

    override fun handleMessage(msg: Message) {
        if (msg.what == LISTEN) {
            try {
                if (mSocket == null) {
                    mSocket = DatagramSocket(null)
                    val socketAddress = InetSocketAddress(InetAddress.getByName("0.0.0.0"), mPort)
                    mSocket!!.apply {
                        this.reuseAddress = true
                        this.broadcast = true
                        this.soTimeout = 0
                    }.bind(socketAddress)
                }
                val recvBuf = ByteArray(15000)
                val packet = DatagramPacket(recvBuf, recvBuf.size)
                val socket = mSocket!!
                socket.receive(packet)

                try {
                    val jsonObject = JSONObject(String(packet.data).trim { it <= ' ' })
                    val host = Host(packet.address, jsonObject.getString(Host.JSON_NAME), jsonObject.getString(Host.JSON_FILTER_TEXT))

                    if ((isHostClientToo || !mCurrentIps.contains(host.hostAddress)) && hostMatchesFilter(host.filterText.trim { it <= ' ' })) {
                        var handler = mHostHandlerMap[host]
                        if (handler == null) {
                            handler = StaleHostHandler(host, mHostHandlerMap, mListener)
                            synchronized(this@UdpBroadcastListeningHandler) { mHostHandlerMap.put(host, handler) }
                            Handler(Looper.getMainLooper()).post { mListener?.onHostsUpdate(mHostHandlerMap.keys) }
                        } else if (hostNameChanged(host, mHostHandlerMap)) {
                            Handler(Looper.getMainLooper()).post { mListener?.onHostsUpdate(mHostHandlerMap.keys) }
                        }
                        handler.removeMessages(StaleHostHandler.STALE_HOST)
                        handler.sendEmptyMessageDelayed(StaleHostHandler.STALE_HOST, mStaleTimeout)
                    }
                } catch (ignored: JSONException) { }
            } catch (e: SocketException) {

                e.printStackTrace()
                mSocket?.close()
                mSocket = null

            } catch (e: UnknownHostException) {

                e.printStackTrace()
                mSocket?.close()
                mSocket = null

            } catch (e: IOException) {
                e.printStackTrace()
                if (mListener != null) {
                    Handler(Looper.getMainLooper()).post { mListener!!.onReceiveFailed() }
                }
            }
            if (looper.thread.isAlive) {
                sendEmptyMessage(LISTEN)
            }
        }
    }

    private fun hostNameChanged(updatedHost: Host, hostHandlerMap: ArrayMap<Host, StaleHostHandler>): Boolean {
        for (host: Host in hostHandlerMap.keys) {
            if (updatedHost == host && updatedHost.name != host.name) {
                hostHandlerMap[updatedHost] = hostHandlerMap.remove(host)
                return true
            }
        }
        return false
    }

    private fun hostMatchesFilter(filterText: String): Boolean {
        return filterText.matches(mRegex)
    }

    private fun stop() {
        for (handler: StaleHostHandler? in mHostHandlerMap.values) {
            handler!!.removeMessages(StaleHostHandler.STALE_HOST)
        }
        mListener = null
        mSocket?.close()
        mSocket = null
        removeMessages(LISTEN)
        looper.quitSafely()
    }

    companion object {
        private const val LISTEN = 5678
        private var handler: UdpBroadcastListeningHandler? = null

        fun startBroadcastListening(hostHandlerMap: ArrayMap<Host, StaleHostHandler>,
                                    currentHostIps: Set<String>,
                                    isHostClientToo: Boolean,
                                    staleTimeout: Long,
                                    port: Int,
                                    regex: Regex) {
            val handlerThread: HandlerThread = object : HandlerThread("ServerService") {
                override fun onLooperPrepared() {
                    handler = UdpBroadcastListeningHandler(looper)
                    handler!!.apply {
                        this.mHostHandlerMap = hostHandlerMap
                        this.mCurrentIps = currentHostIps
                        this.isHostClientToo = isHostClientToo
                        this.mStaleTimeout = staleTimeout
                        this.mPort = port
                        this.mRegex = regex
                    }.sendEmptyMessage(LISTEN)
                }
            }
            handlerThread.start()
        }

        fun setListener(listener: UdpBroadcastListener) {
            handler?.updateListenersTo(listener)
        }

        fun stopListeningForBroadcasts() {
            handler?.stop()
            handler = null
        }
    }
}